plugin/sampling/strategystore/adaptive/strategy_store.go

```diff
@@ -15,32 +15,83 @@
+	storage             samplingstore.Store
+	logger              *zap.Logger
+
+	// probabilities contains the latest calculated sampling probabilities for service operations.
+	probabilities model.ServiceOperationProbabilities
+
+	// strategyResponses is the cache of the sampling strategies for every service, in protobuf format.
+	strategyResponses map[string]*api_v2.SamplingStrategyResponse
+
+	// followerRefreshInterval determines how often the follower processor updates its probabilities.
+	// Given only the leader writes probabilities, the followers need to fetch the probabilities into
+	// cache.
+	followerRefreshInterval time.Duration
+
+	shutdown   chan struct{}
+	bgFinished sync.WaitGroup
+}
+
 // NewStrategyStore creates a strategy store that holds adaptive sampling strategies.
-func NewStrategyStore(options Options, metricsFactory metrics.Factory, logger *zap.Logger, lock distributedlock.Lock, store samplingstore.Store) (*Processor, error) {
-	hostname, err := hostname.AsIdentifier()
-	if err != nil {
-		return nil, err
+func NewStrategyStore(options Options, logger *zap.Logger, participant leaderelection.ElectionParticipant, store samplingstore.Store) (*StrategyStore, error) {
+	return &StrategyStore{
+		Options:                 options,
+		storage:                 store,
+		probabilities:           make(model.ServiceOperationProbabilities),
+		strategyResponses:       make(map[string]*api_v2.SamplingStrategyResponse),
+		logger:                  logger,
+		electionParticipant:     participant,
+		followerRefreshInterval: defaultFollowerProbabilityInterval,
+	}, nil
+}
+
+// Start initializes and starts the sampling service which regularly loads sampling probabilities and generates strategies.
+func (ss *StrategyStore) Start() error {
+	ss.logger.Info("starting adaptive sampling service")
+	if err := ss.electionParticipant.Start(); err != nil {
+		return err
 	}
-	logger.Info("Using unique participantName in adaptive sampling", zap.String("participantName", hostname))
-
-	participant := leaderelection.NewElectionParticipant(lock, defaultResourceName, leaderelection.ElectionParticipantOptions{
-		FollowerLeaseRefreshInterval: options.FollowerLeaseRefreshInterval,
-		LeaderLeaseRefreshInterval:   options.LeaderLeaseRefreshInterval,
-		Logger:                       logger,
-	})
-	p, err := newProcessor(options, hostname, store, participant, metricsFactory, logger)
-	if err != nil {
-		return nil, err
+	ss.shutdown = make(chan struct{})
+	ss.loadProbabilities()
+	ss.generateStrategyResponses()
+	ss.runBackground(ss.runUpdateProbabilitiesLoop)
+	return nil
+}
+
+// Close stops the service from loading probabilities and generating strategies.
+func (ss *StrategyStore) Close() error {
+	ss.logger.Info("stopping adaptive sampling service")
+	err := ss.electionParticipant.Close()
+	if ss.shutdown != nil {
+		close(ss.shutdown)
 	}
+	ss.bgFinished.Wait()
+	return err
+}
 
-	return p, nil
+func (ss *StrategyStore) runBackground(f func()) {
```

Comment:
nit: since this is not used more than once, it's more readable to just inline it in L76